### PR TITLE
RblString comparison primitives

### PR DESCRIPTION
**Pull request by kjmcqueen**
https://rchain.atlassian.net/browse/ROS-253
https://rchain.atlassian.net/browse/ROS-306

RblString comparison primitives. This includes both case sensitive and case insensitive primitives.

I know this is a large PR, and I'm sorry, but the primitives themselves are all small and very similar.

### CODE REVIEW

**Comment by kirkwood** [comment]
Possibly dumb, definitely untested, Scala question:
```Scala
case class strBinOp(name: String, f: (RblString, RblString) => Boolean) extends Prim {
  override val minArgs: Int = 2
  override val maxArgs: Int = 2

  @checkTypeMismatch[RblString]
  @checkArgumentMismatch
  override def fn(ctxt: Ctxt): Either[PrimError, RblBool] = {
    val elem = ctxt.argvec.elem
    val s1   = elem(0).asInstanceOf[RblString]
    val s2   = elem(1).asInstanceOf[RblString]
    Right(RblBool(f(s1, s2)))
  }
}

object rblstring {

  object stringEq  extends strBinOp("string=", (_ == _))
  object stringNEq extends strBinOp("string!=", (_ != _))

  /* ... */
}
```
Is something like that not indicated?

**Comment by kjmcqueen** [comment]
@kirkwood Not a bad idea.  I'll give it a whirl and see if it works.  I've got the tests written, so it won't be hard to find out if it works.

File: roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    }
+  }
+
+  /** Case Insensitive conparisons */
+  object stringCiEq extends Prim {
+    override val name: String = "string-ci="
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[RblString] // Args must be strings
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala]: See comment about `string=`.

File: roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+      Right(RblBool(s1.value.toLowerCase == s2.value.toLowerCase))
+    }
+  }
+
+  object stringCiNEq extends Prim {
+    override val name: String = "string-ci!="
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[RblString] // Args must be strings
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala]: See comment about `string=`.

File: roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+      Right(RblBool(s1.value.toLowerCase != s2.value.toLowerCase))
+    }
+  }
+
+  object stringCiLess extends Prim {
+    override val name: String = "string-ci<"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[RblString] // Args must be strings
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala]: See comment about `string=`.

File: roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+      Right(RblBool(s1.value.toLowerCase < s2.value.toLowerCase))
+    }
+  }
+
+  object stringCiLEq extends Prim {
+    override val name: String = "string-ci<="
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[RblString] // Args must be strings
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala]: See comment about `string=`.

File: roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+      Right(RblBool(s1.value.toLowerCase <= s2.value.toLowerCase))
+    }
+  }
+
+  object stringCiGtr extends Prim {
+    override val name: String = "string-ci>"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[RblString] // Args must be strings
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala]: See comment about `string=`.

File: roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, RblBool] = {
+      val elem = ctxt.argvec.elem
+      val s1   = elem(0).asInstanceOf[RblString]
+      val s2   = elem(1).asInstanceOf[RblString]
+      Right(RblBool(s1.value.toLowerCase > s2.value.toLowerCase))
+    }
+  }
+
+  object stringCiGEq extends Prim {
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala]: `stringCiGEQ`.

File: roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    override val name: String = "string-ci<="
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala]: `>=`.

File: roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[RblString] // Args must be strings
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala]: See comment about `string=`.

File: roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+object rblstring {
+
+  /** Case sensitive comparisons */
+  object stringEq extends Prim {
+    override val name: String = "string="
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[RblString] // Args must be strings
```
[the remark this answers is not in the document]
Comment (kjmcqueen) [reply roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala]: Correct me if I'm wrong, but these are all RblString comparison primitives, and require both arguments to be RblStrings. Am I missing something?

File: roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala (hunk 7 above)
Comment (kjmcqueen) [reply roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala]: I missed one.  Thank you.
Done.

File: roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (kjmcqueen) [reply roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala]: Oh, I see what you mean now. I tested (= "foo" 3) and got #f.  And indeed, checking RBLstring.cc that is the way it is written. Grrrr.  That is not the way I would have written it. Okay, I'll work on this.

File: roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    override val maxArgs: Int = 2
+
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, RblBool] = {
+      val elem = ctxt.argvec.elem
+
+      checkRblStringFirst(elem(0)).flatMap { s1 =>
+        checkRblStringSecond(elem(1)) match {
+          case Right(s2) => Right(RblBool(f(s1.value, s2.value)))
+          case Left(res) => Right(res)
```
Comment (tymm) [inline roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala]: `case Left(_) => Right(RblBool(false))` simplifies it a little more I think.

Comment (kjmcqueen) [reply roscala/src/main/scala/coop/rchain/rosette/prim/rblstring.scala]: I think I like my way better, but have it your way.

Done.

**Pull request merged**